src/sysupdate/sysupdated.c

```diff
@@ -0,0 +1,1498 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "sd-bus.h"
+
+#include "bus-get-properties.h"
+#include "bus-label.h"
+#include "bus-log-control-api.h"
+#include "bus-polkit.h"
+#include "common-signal.h"
+#include "discover-image.h"
+#include "env-util.h"
+#include "fd-util.h"
+#include "fileio.h"
+#include "json.h"
+#include "main-func.h"
+#include "memfd-util.h"
+#include "missing_capability.h"
+#include "mkdir-label.h"
+#include "os-util.h"
+#include "process-util.h"
+#include "service-util.h"
+#include "signal-util.h"
+#include "socket-util.h"
+#include "string-table.h"
+#include "user-util.h"
+
+typedef struct Manager {
+        sd_event *event;
+        sd_bus *bus;
+
+        Hashmap *targets;
+
+        uint32_t current_job_id;
+        Hashmap *jobs;
+
+        Hashmap *polkit_registry;
+
+        int notify_fd;
+        sd_event_source *notify_event_source;
+} Manager;
+
+typedef enum TargetClass {
+        /* These should try to match ImageClass from src/basic/os-util.h */
+        TARGET_MACHINE = IMAGE_MACHINE,
+        TARGET_PORTABLE = IMAGE_PORTABLE,
+        TARGET_SYSEXT = IMAGE_SYSEXT,
+        TARGET_CONFEXT = IMAGE_CONFEXT,
+
+        /* sysupdate-specific classes */
+        TARGET_HOST,
+        TARGET_COMPONENT,
+
+        _TARGET_CLASS_MAX,
+        _TARGET_CLASS_INVALID = -EINVAL,
+} TargetClass;
+
+typedef struct Target {
+        Manager *manager;
+
+        TargetClass class;
+        char *name;
+        char *path;
+        char *version;
+
+        char *id;
+        ImageType image_type;
+        bool busy;
+} Target;
+
+typedef enum JobType {
+        JOB_LIST,
+        JOB_DESCRIBE,
+        JOB_CHECK_NEW,
+        JOB_UPDATE,
+        JOB_VACUUM,
+        _JOB_TYPE_MAX,
+        _JOB_TYPE_INVALID = -EINVAL,
+} JobType;
+
+typedef struct Job Job;
+
+typedef int (*JobComplete)(sd_bus_message *, const Job *, JsonVariant *, sd_bus_error *);
+
+struct Job {
+        Manager *manager;
+        Target *target;
+
+        uint32_t id;
+        char *object_path;
+
+        JobType type;
+        bool offline;
+        char *version; /* Passed into sysupdate for JOB_DESCRIBE and JOB_UPDATE */
+
+        unsigned progress_percent;
+
+        pid_t pid;
```

Comment:
so, for new code please don't unnecessarily store the pid as separate field. instead usesd_event_source_get_child_process_own(), sd_event_source_send_child_signal() and so on.

we live in a pidfd world now, and sd-event provides enough compat for old pre-pidfd logic, so that we can just pretend we don't have to care.